tests/5.0/target_data/test_target_data_use_device_addr.c

```diff
@@ -0,0 +1,58 @@
+//===-- test_target_data_use_device_addr.c - test of use_device_addr on target data ----===//
+// 
+// OpenMP API Version 5.0 Nov 2018
+// 
+// This file is a test for the use_device_addr when used with the map
+// clause with target data directive. This test uses a scalar and an array of size N 
+// which values are modified on the  device and tested in the host. 
+// List items that appear in a use_device_addr clause have the address 
+// of the corresponding object in the device data environment inside the construct. 
+// This test also tests that address conversions of use_device_addr clauses will 
+// occur as if performed after all variables are mapped according to those map clauses.
+//===----------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1000
+
+int main() {
+  int errors = 0;
+  int host_data = 0, host_array[N];
+  int device_data = 14;
+
+
+  for(int i = 0; i < N; i++)
+    host_array[i] = i;
+  
+  OMPVV_TEST_OFFLOADING;
+
+#pragma omp target data map(to: device_data) use_device_addr(device_data)
+    {
+#pragma omp target map(tofrom: host_data, host_array)
```

Comment:
This testcase does not really use "use_device_addr" – in the "target data" section, device_data (with value 0) has the address of the device copy – but it is not used.
In the "target" section, gets an implicit "firstprivate(device_data)" (a device_data is a scalar) and that firstprivatized variable (with value 0) is then used inside target.

I think you need to do something along the lines of:
```
device_data = 42;
#pragma omp target data map(to: device_data) use_device_addr(device_data)
{
  intptr_t ptr_addr = (intptr_t) &device_data;

  #pragma omp target map(to: device_data) map(tofrom:errors)
  {
      // has implicit firstprivate(ptr_addr); map needed to override implicit firstprivate(device_data)
    if ((intptr_t) &device_data != ptr_addr) ++errors;
  }

  #pragma omp target  map(tofrom:errors) // implicit: firstprivate(ptr_addr); device_data still there due to "target data"
  {
    if (*(int*) ptr_addr != 42) ++errors;
  }
```